stellargraph/connector/neo4j/sampler.py

```diff
@@ -122,66 +131,36 @@ def run(
-                UNWIND {{node_id_list}} AS node_id
-
-                // for each node id in every row, collect the random list of its neighbors.
-                CALL apoc.cypher.run(
-
-                    'MATCH(cur_node) WHERE id(cur_node) = {{node_id}}
-
-                    // find the neighbors
-                    MATCH (cur_node){direction[sampling_direction]}(neighbors)
-
-                    // put all ids into a list
-                    WITH CASE collect(id(neighbors)) WHEN [] THEN [null] ELSE collect(id(neighbors)) END AS in_neighbors_list
-
-                    // pick random nodes with replacement
-                    WITH apoc.coll.randomItems(in_neighbors_list, {{num_samples}}, True) AS in_samples_list
-
-                    RETURN in_samples_list',
-                    {{ node_id: node_id, num_samples: {{num_samples}}  }}) YIELD value
-
-                RETURN apoc.coll.flatten(collect(value.in_samples_list)) as next_samples
-            """
-
         self._check_neighbourhood_sizes(in_size, out_size)
         self._check_common_parameters(nodes, n, len(in_size), seed)
 
-        samples = [[head_node for head_node in nodes for _ in range(n)]]
-
-        tree_cur_index = 0
-        tree_end_index = 0
+        head_nodes = [head_node for head_node in nodes for _ in range(n)]
+        hops = [[head_nodes]]
 
-        in_sample_query = bfs_neighbor_query(sampling_direction="IN")
-        out_sample_query = bfs_neighbor_query(sampling_direction="OUT")
+        in_sample_query = _bfs_neighbor_query(sampling_direction="IN")
+        out_sample_query = _bfs_neighbor_query(sampling_direction="OUT")
 
         for in_num, out_num in zip(in_size, out_size):
-            # BFS traversal in the flattened binary tree
-            while tree_cur_index <= tree_end_index:
-                cur_nodes = samples[tree_cur_index]
+            last_hop = hops[-1]
+            this_hop = []
+            for cur_nodes in last_hop:
                 # get in-neighbor nodes
                 neighbor_records = neo4j_graphdb.run(
                     in_sample_query,
                     parameters={"node_id_list": cur_nodes, "num_samples": in_num},
                 )
-                samples.append(neighbor_records.data()[0]["next_samples"])
+                this_hop.append(neighbor_records.data()[0]["next_samples"])
+
                 # get out-neighbor nodes
                 neighbor_records = neo4j_graphdb.run(
                     out_sample_query,
                     parameters={"node_id_list": cur_nodes, "num_samples": out_num},
                 )
-                samples.append(neighbor_records.data()[0]["next_samples"])
+                this_hop.append(neighbor_records.data()[0]["next_samples"])
 
-                tree_cur_index += 1
-            # go to next binary tree depth
-            tree_end_index = tree_end_index * 2 + 2
+            hops.append(this_hop)
 
-        return samples
+        return list(it.chain(*hops))
```

Comment:
I personally find this clearer as (and it's slightly faster):

```suggestion
        return sum(hops, [])
```